text/0034-ternary-hash.md

```diff
@@ -0,0 +1,472 @@
+
+```
+if J_b > to_bin([0, 1, 1, ..., 1]):
+    J_b <- J_b - to_bin([1, 0, 0, ..., 0])
+
+if J_b < (to_bin([0, 0, 0, ..., 0]) - to_bin([0, 1, 1, ..., 1])):
+    J_b <- J_b + to_bin([1, 0, 0, ..., 0])
+```
+
+##### Kerl updates the inner state by applying logical not
+
+The upstream keccak implementation uses a complicated permutation to update the inner state of the sponge construction
+after a hash was squeezed from it. `Kerl` opts to apply logical not, `!`, to the bytes squeezed from `keccak`, and
+updating `keccak`'s inner state with these.
+
+#### Design and implementation
+
+The main goal of the implementation was to ensure that representation of the integer was cast into types. To that end,
+the following ternary types are defined as wrappers around `TritBuf` (read `T243` the same way you would think of `u32`
+or `i64`):
+
+```rust
+struct T242<T: Trit> {
+    inner: TritBuf<T1B1Buf<T>>,
+}
+
+struct T243<T: Trit> {
+    inner: TritBuf<T1B1Buf<T>>,
+}
+```
+
+where `TritBuf`, `T1B1Buf`, and `Trit` are types and traits defined in the `bee-ternary` crate. The bound `Trit` ensures
+that the structs only contain ternary buffers with `Btrit` (for balancer ternary), and `Utrit` (for unbalanced ternary).
+Methods on `T242` and `T243` assume that the most significant trit is always in the last position (think “little
+endian”).
+
+For the binary representation of the integer, the types `I384` (“signed” integer akin to `i64`), and `U384` (“unsigned”
+akin to `u64`) are defined:
+
+```rust
+struct I384<E, T> {
+    inner: T,
+    _phantom: PhantomData<E>.
+}
+
+struct U384<E, T> {
+    inner: T,
+    _phantom: PhantomData<E>.
+}
+```
+
+`inner: T` for encoding the inner fixed-size arrays used for storing either bytes, `u8`, or integers, `u32`:
+
+```rust
+type U8Repr = [u8; 48];
+type U32Repr = [u32; 12];
+```
+
+The phantom type `E` is used to encode endianness, `BigEndian` and `LittleEndian`. These are just used as marker types
+without any methods.
+
+```rust
+struct BigEndian {}
+struct LittleEndian {}
+
+trait EndianType {}
+
+impl EndianType for BigEndian {}
+impl EndianType for LittleEndian {}
+```
+
+The underlying arrays and endianness are important, because `keccak` expects bytes as input, and because `Kerl` made the
+choice to revert the order of the integers in binary big int.
+
+To understand the implementation, the most important methods are:
+
+```rust
+T242<Btrit>::from_i384_ignoring_mst
+T243<Utrit>::from_u384
+I384<LittleEndian, U32Repr>::from_t242
+I384<LittleEndian, U32Repr>::try_from_t242
+U384<LittleEndian, U32Repr>::add_inplace
+U384<LittleEndian, U32Repr>::add_digit_inplace
+U384<LittleEndian, U32Repr>::sub_inplace
+U384<LittleEndian, U32Repr>::from_t242
+U384<LittleEndian, U32Repr>::try_from_t243
+```
+
+# Drawbacks
+
++ All hash functions, no matter if they can fail or not, have to implement `Error`;
++ There are a lot of types. Is it important to encode that the most significant trit is `0` by having a `T242`?;
```

Comment:
regarding: "is it important to encode...." Isn't that an "unresolved question"? IMHO: "there are a lot of types" is not in-and-of-itself a drawback.